olp-cpp-sdk-core/src/client/OlpClient.cpp

```diff
@@ -545,95 +487,67 @@ HttpResponse OlpClient::OlpClientImpl::CallApi(
                         "Network request handler is empty.");
   }
 
-  const auto& retry_settings = settings_.retry_settings;
-  auto network_settings =
-      http::NetworkSettings()
-          .WithTransferTimeout(retry_settings.timeout)
-          .WithConnectionTimeout(retry_settings.timeout)
-          .WithProxySettings(
-              settings_.proxy_settings.value_or(http::NetworkProxySettings()));
-
-  auto network_request = http::NetworkRequest(
-      utils::Url::Construct(GetBaseUrl(), path, query_params));
-
-  network_request.WithVerb(GetHttpVerb(method))
-      .WithBody(std::move(post_body))
-      .WithSettings(std::move(network_settings));
-
-  for (const auto& header : default_headers_) {
-    network_request.WithHeader(header.first, header.second);
-  }
-
-  std::string custom_user_agent;
-  for (const auto& header : header_params) {
-    // Merge all User-Agent headers into one header.
-    // This is required for (at least) iOS network implementation,
-    // which uses headers dictionary without any duplicates.
-    // User agents entries are usually separated by a whitespace, e.g.
-    // Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:47.0) Firefox/47.0
-    if (CaseInsensitiveCompare(header.first, http::kUserAgentHeader)) {
-      custom_user_agent += header.second + std::string(" ");
-    } else {
-      network_request.WithHeader(header.first, header.second);
-    }
-  }
+  auto network_request = CreateRequest(path, method, query_params,
```

Comment:
One though. Is it not the same if you call async CallApi() from the sync CallApi()?
Something like:

```cpp
HttpResponse CallApi(std::string path, std::string method,
                       ParametersType query_params,
                       ParametersType header_params, ParametersType form_params,
                       RequestBodyType post_body, std::string content_type,
                       CancellationContext context) const {
   auto callback = [&](HttpResponse http_response) {
        response_data->response = std::move(http_response);
        response_data->condition.Notify();
   };
  
   context.ExecuteOrCancelled([&](){return CallApi(path, method, ..., callback)}, 
                                                    [&](){ ... });
   
     // Wait for response_data->condition and return response.
}
```

Reply:
this is can be the most elegant solution. But wouldn't we need to add condition.Notify() into cancellation tokens created inside async CallApi?

Reply:
No as it is added to the CancellationContext. If you look at CancellationContext::ExecuteOrCancelled() then you see that the CancellationToken of the sub-operation is remembered:

```cpp